src/mlpack/bindings/cli/add_to_po.hpp

```diff
@@ -14,95 +13,171 @@
 #define MLPACK_BINDINGS_CLI_ADD_TO_PO_HPP
 
 #include <mlpack/core/util/param_data.hpp>
-#include <boost/program_options.hpp>
 #include <mlpack/core/util/is_std_vector.hpp>
 #include "map_parameter_name.hpp"
 
+#include "third_party/CLI/CLI11.hpp"
+
 namespace mlpack {
 namespace bindings {
 namespace cli {
 
 /**
- * Add a non-vector option to boost::program_options.
- *
- * @param boostName The name of the option to add to boost::program_options.
+ * Add a tuple option to CLI11.
+ * 
+ * @param cliName The name of the option to add to CLI11.
  * @param descr Description string for parameter.
  * @param desc Options description to add parameter to.
  */
 template<typename T>
-void AddToPO(const std::string& boostName,
-             const std::string& descr,
-             boost::program_options::options_description& desc,
-             const typename boost::disable_if<util::IsStdVector<T>>::type* = 0,
-             const typename boost::disable_if<std::is_same<T, bool>>::type* = 0)
+void AddToPO(const std::string& cliName,
+             util::ParamData& param,
+             CLI::App& app,
+             const typename boost::disable_if<std::is_same<T, bool>>::type* = 0,
+             const typename boost::disable_if<arma::is_arma_type<T>>::type* = 0,
+             const typename boost::disable_if<data::HasSerialize<T>>::type* = 0,
+             const typename boost::enable_if<std::is_same<T,
+                std::tuple<mlpack::data::DatasetInfo, arma::mat>>>::type* = 0)
 {
-  desc.add_options()(boostName.c_str(), boost::program_options::value<T>(),
-      descr.c_str());
+  app.add_option_function<T>(cliName.c_str(),
+      [&param](const T& value)
+      {
+        param.value = value;
+        param.wasPassed = true;
+      },
+      param.desc.c_str());
```

Comment:
So, let's assume you've made the change below so that `T` is passed directly from the bottom overload to here.  This overload applies when `T` is `std::tuple<mlpack::data::DatasetInfo, arma::mat>`.  But like I explained on the call, the `boost::any` of `param.value` is not going to hold a `std::tuple<mlpack::data::DatasetInfo, arma::mat>`; we also have to hold the filename, so it will *actually* hold a `std::tuple<std::tuple<mlpack::data::DatasetInfo, arma::mat>, std::string>`, where that last `std::string` is holding the filename of the matrix.  Therefore, we have to account for that in this overload.  Thus, the lambda here should be:

```
app.add_option_function<T>(cliName.c_str(),
    [&param](const T& value)
    {
      typedef std::tuple<T, typename ParameterType<T>::type> TupleType;
      TupleType& tuple = *boost::any_cast<TupleType>(&param.value);
      std::get<1>(tuple) = boost::any_cast<std::string>(value);
      param.wasPassed = true;
    },
    param.desc.c_str());
```

Note that the inner implementation I am using there is directly taken from `set_param.hpp`. :+1: